asf_core_data/pipeline/preprocessing/feature_engineering.py

```diff
@@ -43,6 +44,101 @@
 # ----------------------------------------------------------------------------------
 
 
+def prepare_address_data(address_1: str, address_2: str, postcode: str) -> str:
+    """
+    Prepares address and postcode information by removing removing punctuation,
+    removing leading and trailing whitespace and standardising postcodes.
+
+    Args:
+        address_1: first line of address (e.g. "Flat 1")
+        address_2: second line of address (e.g. "ABC road")
+        postcode: full postcode (e.g. "AB2 Y8X")
+
+
+    Returns:
+        A string with processed address_1, address_2 and postcode information,
+        e.g. "flat 1 abc road_AB2Y8X"
+    """
+    address_1 = remove_punctuation(address_1).lower().strip()
+    address_2 = remove_punctuation(address_2).lower().strip()
+    postcode = postcode.upper().replace(" ", "")
+
+    address_info = (address_1 + " " + address_2).strip()
+
+    return "_".join([address_info, postcode])
+
+
+def concatenate_building_with_address_info(
+    building_reference: str, address_info: str
+) -> str:
+    """
+    Concatenates building reference number with address information.
+
+    Args:
+        building_reference: building reference number, e.g. 1234
+        address_info: address information, e.g. "flat 1 abc road_AB2Y8X"
+
+    Returns:
+        A string with concatenated building reference and address info (e.g. "1234_flat 1 abc road_AB2Y8X")
+    """
+    return str(building_reference) + "_" + address_info
+
+
+def enhance_uprn(df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Overrall, I quite like this PR as the changes are specific, and the logic is well commented.

I would extract parts of this out as it's own function (i.e. mapping creation) on the asf_daps side. But I'm happy to merge this PR here.

Reply:
thanks for review @sqr00t ! Agree with further refactoring the `enhance_uprn()` function, but if you'll do it on asf_daps side, i'll leave it like this there.